lib/jekyll/converters/markdown/kramdown_parser.rb

```diff
@@ -1,33 +1,95 @@
+# Frozen-string-literal: true
+# Encoding: utf-8
+
 module Jekyll
   module Converters
     class Markdown
       class KramdownParser
+        CoderayDefaults = {
+          "bold_every" => 10,
+          "css" => "style",
+          "line_numbers" => "inline",
+          "line_number_start" => 1,
+          "wrap" => "div",
+          "tab_width" => 4
+        }
+
         def initialize(config)
-          require 'kramdown'
-          @config = config
-          # If kramdown supported highlighter enabled, use that
-          highlighter = @config['highlighter']
-          if highlighter == 'rouge' || highlighter == 'coderay'
-            @config['kramdown']['syntax_highlighter'] ||= highlighter
+          @jekyll_config = config
+          @config = @jekyll_config["kramdown"] ||= {}
+          load_kramdown
+          setup
+        end
+
+        # Setup and normalize the configuration:
+        #   * Create Kramdown if it doesn't exist.
+        #   * Set syntax_highlighter, detecting enable_coderay and merging highlighter if none.
+        #   * Merge kramdown[coderay] into syntax_highlighter_opts stripping coderay_.
+        #   * Make sure `syntax_highlighter_opts` exists.
+
+        def setup
+          @config["syntax_highlighter"] ||= highlighter
+          @config["syntax_highlighter_opts"] ||= {}
+          @config["coderay"] ||= {} # XXX: Legacy.
+
+          if highlighter == "coderay"
+            ogconfg = @config["syntax_highlighter_opts"]
+            coderay = Jekyll::Utils.deep_merge_hashes(CoderayDefaults, @config["coderay"]) # XXX: Legacy
+            @config["syntax_highlighter_opts"] = strip_coderay(Jekyll::Utils.deep_merge_hashes(
+              ogconfg, coderay
+            ))
           end
-        rescue LoadError
-          STDERR.puts 'You are missing a library required for Markdown. Please run:'
-          STDERR.puts '  $ [sudo] gem install kramdown'
-          raise Errors::FatalException.new("Missing dependency: kramdown")
+
+          @config["syntax_highlighter_opts"] = @config[
+            "syntax_highlighter_opts"
+          ]
         end
 
         def convert(content)
-          # Check for use of coderay
-          if @config['kramdown']['enable_coderay']
-            %w[wrap line_numbers line_numbers_start tab_width bold_every css default_lang].each do |opt|
-              key = "coderay_#{opt}"
-              @config['kramdown'][key] = @config['kramdown']['coderay'][key] unless @config['kramdown'].key?(key)
+          Kramdown::Document.new(content, @config).to_html
+        end
+
+        private
+        def load_kramdown
```

Comment:
this can be replaced by a call to `Jekyll::External.gracefully_require 'kramdown'` which has an error message and such built into it